src/python/pants/util/fileutil.py

```diff
@@ -22,6 +23,17 @@ def atomic_copy(src, dst):
     os.rename(tmp_dst.name, dst)
 
 
+@contextmanager
+def safe_temp_edit(filename):
+  """Safely modify a file within context that automatically reverts any changes afterwards"""
```

Comment:
Mutating files in the buildroot like this is effectively a non-starter. For one thing, if you crash, this will not get cleaned up. But also, this will cause `pantsd` to churn unnecessarily for file invalidations (and will break in the presence of concurrency).

Is it possible to run the relevant install command with an explicit flag that tells it which `package.json` file to use? Something like `yarn install --package-json=${my_tmp_file}`?

Reply:
Unfortunately the path to the package.json is heavily relied on the node package manager tools as the package.json serves as the `root` location that all other relative paths relies on. I don't think this would be supported any time. The mutation for buildroot would only occur for the `node-install` command which is used to mutate the `yarn.lock` file (such as adding a new dependency). Both the `package.json` and the `yarn.lock` files are required to be checked in. There is potential case for lossy, as mentioned above a non-backwards compatibile solution could be implemented in the future to better support pants-specific dependencies. Hopefully any catastrophic loss of `package.json` can be recovered by the code versioning system.

Reply:
Can you put these caveats in comments both 1) at the callsite for this method, 2) in the method doc to indicate that it is actually mutating the file in place.